google-auth-provider/main.go

```diff
@@ -0,0 +1,189 @@
+	oauthProxyOpts.ReverseProxy = true
+	if opts.AuthEmailDomains != "" {
+		oauthProxyOpts.EmailDomains = strings.Split(opts.AuthEmailDomains, ",")
+	}
+
+	if err = validation.Validate(oauthProxyOpts); err != nil {
+		fmt.Printf("failed to validate options: %v\n", err)
+		os.Exit(1)
+	}
+
+	oauthProxy, err := oauth2proxy.NewOAuthProxy(oauthProxyOpts, oauth2proxy.NewValidator(oauthProxyOpts.EmailDomains, oauthProxyOpts.AuthenticatedEmailsFile))
+	if err != nil {
+		fmt.Printf("failed to create oauth2 proxy: %v\n", err)
+		os.Exit(1)
+	}
+
+	port := os.Getenv("PORT")
+	if port == "" {
+		port = "9999"
+	}
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/{$}", func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte(fmt.Sprintf("http://127.0.0.1:%s", port)))
+	})
+	mux.HandleFunc("/obot-get-state", obotGetState(oauthProxy))
+	mux.HandleFunc("/obot-get-icon-url", obotGetIconURL)
+	mux.HandleFunc("/", oauthProxy.ServeHTTP)
+
+	fmt.Printf("listening on 127.0.0.1:%s\n", port)
+	if err := http.ListenAndServe("127.0.0.1:"+port, mux); !errors.Is(err, http.ErrServerClosed) {
+		fmt.Printf("failed to listen and serve: %v\n", err)
+		os.Exit(1)
+	}
+}
+
+type SerializableRequest struct {
+	Method string              `json:"method"`
+	URL    string              `json:"url"`
+	Header map[string][]string `json:"header"`
+}
+
+type SerializableState struct {
+	ExpiresOn         *time.Time `json:"expiresOn"`
+	AccessToken       string     `json:"accessToken"`
+	PreferredUsername string     `json:"preferredUsername"`
+	User              string     `json:"user"`
+	Email             string     `json:"email"`
+}
+
+func obotGetState(p *oauth2proxy.OAuthProxy) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var sr SerializableRequest
+		if err := json.NewDecoder(r.Body).Decode(&sr); err != nil {
+			http.Error(w, fmt.Sprintf("failed to decode request body: %v", err), http.StatusBadRequest)
+			return
+		}
+
+		reqObj, err := http.NewRequest(sr.Method, sr.URL, nil)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("failed to create request object: %v", err), http.StatusBadRequest)
+			return
+		}
+
+		reqObj.Header = sr.Header
+
+		state, err := p.LoadCookiedSession(reqObj)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("failed to load cookied session: %v", err), http.StatusBadRequest)
+			return
+		}
+
+		if state == nil {
+			http.Error(w, "state is nil", http.StatusInternalServerError)
+			return
+		}
+
+		ss := SerializableState{
+			ExpiresOn:         state.ExpiresOn,
+			AccessToken:       state.AccessToken,
+			PreferredUsername: state.PreferredUsername,
+			User:              state.User,
+			Email:             state.Email,
+		}
+
+		if err := json.NewEncoder(w).Encode(ss); err != nil {
+			http.Error(w, fmt.Sprintf("failed to encode state: %v", err), http.StatusInternalServerError)
+			return
+		}
+	}
+}
+
+func obotGetIconURL(w http.ResponseWriter, r *http.Request) {
```

Comment:
Same comment here about importing this from a shared package.

Reply:
Done

Reply:
Doesn't look like this was done.

Reply:
Oh I didn't pay close enough attention to this one lol. Sorry. I would like to keep this out of the common package, since the function is a bit different for our current two providers.